File: server/internal/middleware/ratelimit.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package middleware

import (
	"net"
	"strings"
	"sync"
	"time"

	"github.com/hangtiancheng/yukino.go/yukino_http"
)

// RateLimit throttles an endpoint per client IP: at most max requests within
// each sliding window. It guards the unauthenticated endpoints (login,
// register, password reset) against credential-stuffing and brute-force
// attempts.
func RateLimit(max int, window time.Duration) yukino_http.Middleware {
	var mu sync.Mutex
	hits := make(map[string][]time.Time)
	return func(ctx *yukino_http.Context, next func()) {
		ip := clientIP(ctx)
		now := time.Now()
		mu.Lock()
		recent := hits[ip][:0]
		for _, t := range hits[ip] {
			if now.Sub(t) < window {
				recent = append(recent, t)
			}
		}
		if len(recent) >= max {
			hits[ip] = recent
			mu.Unlock()
			ctx.Status = 200
			ctx.JSON(yukino_http.H{"code": 429, "message": "too many requests, slow down"})
			return
		}
		hits[ip] = append(recent, now)
		mu.Unlock()
		next()
	}
}

// clientIP prefers the first X-Forwarded-For entry (set by a trusted proxy)
// and falls back to the connection's remote address.
func clientIP(ctx *yukino_http.Context) string {
	if fwd := ctx.Get("X-Forwarded-For"); fwd != "" {
		if first, _, ok := strings.Cut(fwd, ","); ok {
			return strings.TrimSpace(first)
		}
		return strings.TrimSpace(fwd)
	}
	host, _, err := net.SplitHostPort(ctx.Request.RemoteAddr)
	if err != nil {
		return ctx.Request.RemoteAddr
	}
	return host
}

File: server/src/middleware/ratelimit.ts
import type { Context, MiddlewareHandler } from "hono";
import { rateLimited } from "../common/envelope.js";

function clientIP(c: Context): string {
  const fwd = c.req.header("X-Forwarded-For");
  if (fwd) {
    const first = fwd.split(",")[0];
    if (first) return first.trim();
    return fwd.trim();
  }
  const incoming = c.env.incoming;
  const addr = incoming.socket.remoteAddress;
  return addr ?? "unknown";
}

// Per-IP sliding window, per middleware instance (Go parity: /login and
// /register each get their own bucket).
export const rateLimit = (max: number, windowMs: number): MiddlewareHandler => {
  const hits = new Map<string, number[]>();

  return async (c, next) => {
    const ip = clientIP(c);
    const now = Date.now();
    const recent = (hits.get(ip) ?? []).filter((t) => now - t < windowMs);
    if (recent.length >= max) {
      hits.set(ip, recent);
      return rateLimited(c);
    }
    recent.push(now);
    hits.set(ip, recent);
    await next();
  };
};
